contracts/oracles/OracleManager.sol

```diff
@@ -45,6 +55,24 @@ contract OracleManager is IOracleManager {
     initialEpochStartTimestamp = getEpochStartTimestamp() - epochLength;
   }
 
+  /*╔═════════════════════════════╗
+    ║          LastRoundId        ║
+    ╚═════════════════════════════╝*/
+
+  function setLastRoundId(uint16 _phaseId, uint64 _lastRoundId) external {
+    (uint80 latestId, , , , ) = chainlinkOracle.latestRoundData();
+    require(latestId >> 64 > _phaseId, "incorrect phase change passed");
+    (, , uint256 nextTimestampOnCurrentPhase, , ) = chainlinkOracle.getRoundData((uint80(_phaseId) << 64) | uint80(_lastRoundId + 1));
+    require(nextTimestampOnCurrentPhase == 0, "incorrect phase change passed");
+
+    lastRoundId[_phaseId] = uint80((uint256(_phaseId) << 64) | _lastRoundId);
+
+    (, , uint256 lastUpdateOnPhaseTimestamp, , ) = chainlinkOracle.getRoundData(lastRoundId[_phaseId]);
+
+    // NOTE: this protects against chainlink phases with no price updates
+    require(lastUpdateOnPhaseTimestamp > 0 || _lastRoundId == 0, "incorrect phase change passed");
+  }
```

Comment:
In the rare case that a phase changes prevents us from scanning chainlink prices backwards we can set them here.

This bug wasn't picked up by the audit competition. It is the only 'hard' bug in the code, the rest were more opinion and nitpicking on outdated comments, errors in readmes etc.

As such it is also the most complicated change in this diff. But we wrote some good tests on it (excluded from the diff to keep the code in check).

Here are the two test files:
```solidity
// SPDX-License-Identifier: BUSL-1.1
pragma solidity 0.8.17;

//import "forge-std/console2.sol";

import "./FloatContractsCoordinator.s.sol";
//import "./ChainlinkAggregator.sol";
import "./ChainlinkMockAggregator.sol";

interface AggregatorV2V3InterfaceProposer is AggregatorV2V3Interface
{
    function proposeAggregator(address _aggregator) external;
    function confirmAggregator(address _aggregator) external;
    function aggregator()
        external
        returns (address);
}

/**
   @dev - SoftUpgrade doesn't redeploy a new shifter, it just updates its implementetaiton (for non-breaking changes)
*/
contract ChainlinkPhaseChangeTest is FloatContractsCoordinator {
    address chainlinkProxyAddress = 0xF9680D99D6C9589e2a93a78A04A279e509205945;
    address chainlinkOwner = 0x5A3ca642fdEd17296adDdb53A496ce4f26901596;
    address chainlinkAggregator = 0x4dD6655Ad5ed7C06c882f496E3f42acE5766cb89;

    OffchainAggregatorMock newAggregator;

    uint32 marketIndex;

    uint256 initialPoolLiquidity;

    function setUp() public {
        // vm.startPrank(vm.envAddress("MUMBAI_ALPHA_ADMIN_ADDRESS"));
        // vm.startPrank(0x921c380d6047f09a4E61D29d1Fc054fB52259CA9);
        // vm.selectFork(vm.createFork(vm.envString("MUMBAI_RPC_URL")));
        vm.selectFork(vm.createFork("https://polygon-rpc.com"));
        // vm.selectFork(vm.createFork("https://matic-mumbai.chainstacklabs.com/"));
        // vm.rollFork(36093716);

        gems = constructGems();
        registry = constructRegistry(address(gems), ADMIN);

        marketFactory = new MarketFactory(registry);

        bytes memory name = "defaultTestPaymentToken";
        uint32 identifier = 0;

        PaymentTokenTestnet paymentToken = (new PaymentTokenTestnet());
        paymentToken.initializeWithAdmin(ADMIN, string.concat("Test Payment Token ", string(name)), string.concat("TPT", Strings.toString(identifier)));
        deal(address(paymentToken), ADMIN, 1e22);

    initialPoolLiquidity = 100e18;

    MarketFactory.PoolLeverage[] memory poolLeverages = new MarketFactory.PoolLeverage[](5);
    poolLeverages[0] = MarketFactory.PoolLeverage(1e18, IMarketCommon.PoolType.FLOAT, 0);
    poolLeverages[1] = MarketFactory.PoolLeverage(1e18, IMarketCommon.PoolType.LONG, 0);
    poolLeverages[2] = MarketFactory.PoolLeverage(2e18, IMarketCommon.PoolType.LONG, 1);
    poolLeverages[3] = MarketFactory.PoolLeverage(1e18, IMarketCommon.PoolType.SHORT, 0);
    poolLeverages[4] = MarketFactory.PoolLeverage(2e18, IMarketCommon.PoolType.SHORT, 1);
    console2.log("hi");

    marketIndex = marketFactory.deployMarketWithPrank(
                                            ADMIN,
                                            initialPoolLiquidity,
                                            poolLeverages,
                                            DEFAULT_FIXED_EPOCH_LENGTH,
                                            DEFAULT_MINIMUM_EXECUTION_WAITING_TIME,
                                            chainlinkProxyAddress,
                                            IERC20(address(paymentToken)),
                                            DEFAULT_MARKET_TYPE
    );
    console2.log("hi2");

        vm.startPrank(chainlinkOwner);

        AggregatorV2V3InterfaceProposer chainlinkProxy = AggregatorV2V3InterfaceProposer(chainlinkProxyAddress);
        (
         uint80 roundId,
         int256 answer,
         uint256 startedAt,
         uint256 updatedAt,
         uint80 answeredInRound
        ) = chainlinkProxy.latestRoundData();

        console2.log("stent roundId", roundId);
        console2.log("stent phaseId", roundId >> 64);
        console2.log("stent aggregator roundId", uint64(roundId));

        newAggregator = new OffchainAggregatorMock();

        chainlinkProxy.proposeAggregator(address(newAggregator));
        chainlinkProxy.confirmAggregator(address(newAggregator));

        newAggregator.setRound(100, block.timestamp);

        (
         roundId,
         answer,
         startedAt,
         updatedAt,
         answeredInRound
        ) = chainlinkProxy.latestRoundData();

        console2.log("stent roundId", roundId);
        console2.log("stent phaseId", roundId >> 64);
        console2.log("stent aggregator roundId", uint64(roundId));

        vm.stopPrank();
    }

    function timeLeftInEpoch() public view returns (uint256 _timeLeftInEpoch) {
        IOracleManager defaultOracleManager = marketFactory.oracleManager(marketIndex);
        uint256 epochStartTimestamp = defaultOracleManager.getEpochStartTimestamp();
        _timeLeftInEpoch = epochStartTimestamp + defaultOracleManager.EPOCH_LENGTH() - block.timestamp;
    }

    function warpToEndOfMewtInNextEpoch() public {
        vm.warp(block.timestamp + timeLeftInEpoch() + marketFactory.oracleManager(marketIndex).MINIMUM_EXECUTION_WAIT_THRESHOLD());
    }

    function testStent() public {
        IMarket market = marketFactory.market(marketIndex);
        AggregatorV2V3InterfaceProposer chainlinkProxy = AggregatorV2V3InterfaceProposer(chainlinkProxyAddress);
        AggregatorV2V3InterfaceMock aggregator = AggregatorV2V3InterfaceMock( chainlinkProxy.aggregator() );
        aggregator.setRound(101, block.timestamp);
        warpToEndOfMewtInNextEpoch();
        aggregator.setRound(102, block.timestamp);

        (
         uint80 roundId,
         int256 answer,
         uint256 startedAt,
         uint256 updatedAt,
         uint80 answeredInRound
        ) = chainlinkProxy.latestRoundData();

        console2.log("stent roundId", roundId);
        console2.log("stent phaseId", roundId >> 64);
        console2.log("stent aggregator roundId", uint64(roundId));

        uint80[] memory missedEpochsOracleRoundIds = new uint80[](1);
        missedEpochsOracleRoundIds[0] = uint80(chainlinkProxy.latestRound());
        market.updateSystemStateUsingValidatedOracleRoundIds(missedEpochsOracleRoundIds);

        aggregator.setRound(103, block.timestamp);
        warpToEndOfMewtInNextEpoch();
        aggregator.setRound(104, block.timestamp);
        uint80 r = uint80(chainlinkProxy.latestRound());
        vm.warp(block.timestamp+100);
        aggregator.setRound(105, block.timestamp);
        vm.warp(block.timestamp+100);
        aggregator.setRound(106, block.timestamp);
        vm.warp(block.timestamp+100);
        aggregator.setRound(107, block.timestamp);
        vm.warp(block.timestamp+100);
        aggregator.setRound(108, block.timestamp);

        (
         roundId,
         answer,
         startedAt,
         updatedAt,
         answeredInRound
        ) = chainlinkProxy.latestRoundData();

        console2.log("stent roundId", roundId);
        console2.log("stent phaseId", roundId >> 64);
        console2.log("stent aggregator roundId", uint64(roundId));

        missedEpochsOracleRoundIds = new uint80[](1);
        missedEpochsOracleRoundIds[0] = uint80(r);
        market.updateSystemStateUsingValidatedOracleRoundIds(missedEpochsOracleRoundIds);

        vm.startPrank(chainlinkOwner);
        newAggregator = new OffchainAggregatorMock();
        chainlinkProxy.proposeAggregator(address(newAggregator));
        chainlinkProxy.confirmAggregator(address(newAggregator));
        vm.stopPrank();

        warpToEndOfMewtInNextEpoch();
        newAggregator.setRound(200, block.timestamp);
        r = uint80(chainlinkProxy.latestRound());

        (
         roundId,
         answer,
         startedAt,
         updatedAt,
         answeredInRound
        ) = chainlinkProxy.latestRoundData();

        marketFactory.oracleManager(marketIndex).setLastRoundId(3, 9);

        console2.log("stent roundId", roundId);
        console2.log("stent phaseId", roundId >> 64);
        console2.log("stent aggregator roundId", uint64(roundId));

        missedEpochsOracleRoundIds = new uint80[](1);
        missedEpochsOracleRoundIds[0] = uint80(r);
        market.updateSystemStateUsingValidatedOracleRoundIds(missedEpochsOracleRoundIds);
    }
}
```
```solidity
pragma solidity 0.8.17;

import "../interfaces/chainlink/AggregatorV3Interface.sol";

interface AccessControllerInterface {
    function hasAccess(address user, bytes calldata data) external view returns (bool);
}

interface AggregatorInterface {
    function latestAnswer() external view returns (int256);
    function latestTimestamp() external view returns (uint256);
    function latestRound() external view returns (uint256);
    function getAnswer(uint256 roundId) external view returns (int256);
    function getTimestamp(uint256 roundId) external view returns (uint256);
}


interface AggregatorV2V3Interface is AggregatorInterface, AggregatorV3Interface
{
}

interface AggregatorV2V3InterfaceMock is AggregatorV2V3Interface
{
    function setRound(int256 answer, uint256 timestamp) external;
}


contract OffchainAggregatorMock is AggregatorV2V3InterfaceMock {

    uint80 currentRoundId;
    int256 currentAnswer;
    uint256 currentTimestamp;

    struct Transmission {
        int256 answer; // 192 bits ought to be enough for anyone
        uint256 timestamp;
    }
        mapping(uint80 /* aggregator round ID */ => Transmission) internal transmissions;

    function setRound(int256 answer, uint256 timestamp) public {
        currentRoundId++;
        currentAnswer = answer;
        currentTimestamp = timestamp;
        Transmission memory t = Transmission(answer, timestamp);
        transmissions[currentRoundId] = t;
    }

    function latestAnswer() external view returns (int256) {
        return currentAnswer;
    }

    function latestTimestamp() external view returns (uint256) {
        return currentTimestamp;
    }
    function latestRound() external view returns (uint256) {
        return uint256(currentRoundId);
    }
    function getAnswer(uint256 roundId) external view returns (int256) {
        return currentAnswer;
    }
    function getTimestamp(uint256 roundId) external view returns (uint256) {
        return currentTimestamp;
    }
    function decimals() external view returns (uint8) {
        return 18;
    }
    function description() external view returns (string memory) {
        return "hi";
    }
    function version() external view returns (uint256) {
        return 1;
    }


    // getRoundData and latestRoundData should both raise "No data present"
    // if they do not have data to report, instead of returning unset values
    // which could be misinterpreted as actual reported values.
    function getRoundData(uint80 _roundId)
        external
        view
        returns (
                 uint80 roundId,
                 int256 answer,
                 uint256 startedAt,
                 uint256 updatedAt,
                 uint80 answeredInRound
        ) {
        Transmission memory t = transmissions[_roundId];
        return (
                _roundId,
                t.answer,
                t.timestamp,
                t.timestamp,
                1
        );
    }

    function latestRoundData()
        external
        view
        returns (
                 uint80 roundId,
                 int256 answer,
                 uint256 startedAt,
                 uint256 updatedAt,
                 uint80 answeredInRound
        ) {
        return (
                currentRoundId,
                currentAnswer,
                currentTimestamp,
                currentTimestamp,
                1
        );
    }
}
```

Thankfully in this unlikely case - if we didn't catch this error all that would happen is the markets would pause until we are able to resolve the phases.